### PR TITLE
Fix CSS escape parsing and redundant restore matching

### PR DESCRIPTION
**Pull request by jrandolf**
This fixes a problem related to parsing CSS escapes and removes some redundant matching done during restoration of strings and parentheses.

Fixed: #30

### CODE REVIEW

File: parsel.ts [hunk 1]
```diff
@@ -431,6 +503,10 @@ export interface Tokens {
 	argument?: string;
 	subtree?: AST;
 	caseSensitive?: 'i';
+	/**
+	 * @internal
```
Comment (LeaVerou) [inline parsel.ts:507]: We still want to document what this is (which was non-obvious to me)

Comment (jrandolf) [reply parsel.ts]: Fixed. (No longer relevant)

**Comment by LeaVerou** [comment]
Hey, thanks for working on this! I was wondering if you could summarize what these changes *do* exactly? Otherwise it requires way more cycles to review than what I have right now, and I don't want it to be delayed.

**Comment by jrandolf** [comment]
Inspired from previous methods, we now replace escapes with a suitable placeholder. In doing so, we don't accidentally parse escaped sequences such as `\"` and `\(`.

It also simplifies any regular expressions that uses escapes sequences such as namespaces since they will be matched as a part of being non-ASCII (due to the placeholder).

The down side is the grammar is not strict; escapes are technically illegal in some places, but it's still correct for the spec which means parsel can parse a superset of the CSS spec.

**Comment by LeaVerou** [comment]
Thank you for the very nice summary. Replacing with a placeholder is kind of a heavyweight solution, I think there are simpler ways to resolve this. I'd probably use a [negative lookbehind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Assertions#:~:text=Negative%20lookbehind%20assertion) in the pseudo regexes, plus a check [here](https://github.com/LeaVerou/parsel/blob/master/parsel.ts#L54). 
Is there an advantage of this approach? (besides Safari < 16 support, which is a temporary problem)

> which means parsel can parse a superset of the CSS spec.

That's actually fine, being permissive is one of Parsel's design principles. 😊

**Comment by jrandolf** [comment]
> Thank you for the very nice summary. Replacing with a placeholder is kind of a heavyweight solution, I think there are simpler ways to resolve this. I'd probably use a [negative lookbehind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Assertions#:~:text=Negative%20lookbehind%20assertion) in the pseudo regexes, plus a check [here](https://github.com/LeaVerou/parsel/blob/master/parsel.ts#L54).
Is there an advantage of this approach? (besides Safari < 16 support, which is a temporary problem)

Negative lookbehind doesn't work because you may need to look back infinitely. For example `\\"` is actually `\\` and `"`. A pathological case is `\\\\\\\\\a`. Look-back won't work there.

This solution is actually quite fast, in fact, it's faster now since the regular expressions don't have to looked at the 'escaped character' branch. Purely because of this fact, parsing manually (character by character) would have worse performance from E2E.

**Review by LeaVerou — APPROVED** [review]
Ok, I'm gonna go ahead and approve this, but I'd strongly recommend you add more code comments to reflect the info in your comments here.

**Pull request merged**